The job `x86_64-gnu-aux` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=6755&view=logs&jobId=ac463cd8-2ed1-5bc0-c9e7-7aad5c967a56) ([raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/6755/logs/1021)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-08-26T13:41:33.2100595Z 306 |         services: &BlobImageResources,
2019-08-26T13:41:33.2100661Z     |                    ^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn BlobImageResources`
2019-08-26T13:41:33.2100725Z 
2019-08-26T13:41:51.2543655Z    Compiling winit v0.16.2
2019-08-26T13:42:19.5339016Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5346290Z    --> webrender/src/frame_builder.rs:111:37
2019-08-26T13:42:19.5350522Z     |
2019-08-26T13:42:19.5354787Z 111 |     #[cfg_attr(feature = "capture", serde(skip))] //TODO
2019-08-26T13:42:19.5362572Z     |
2019-08-26T13:42:19.5367128Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5367128Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5372073Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5374810Z 
2019-08-26T13:42:19.5379329Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5404440Z     |
2019-08-26T13:42:19.5404440Z     |
2019-08-26T13:42:19.5404625Z 216 |     #[cfg_attr(any(feature = "capture", feature = "replay"), serde(skip))]
2019-08-26T13:42:19.5404925Z     |
2019-08-26T13:42:19.5405470Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5405470Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5405676Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5405782Z 
2019-08-26T13:42:19.5411643Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5420254Z     |
2019-08-26T13:42:19.5420254Z     |
2019-08-26T13:42:19.5424669Z 336 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:19.5433096Z     |
2019-08-26T13:42:19.5436823Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5436823Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5441515Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5444441Z 
2019-08-26T13:42:19.5500642Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5502075Z     |
2019-08-26T13:42:19.5502075Z     |
2019-08-26T13:42:19.5502144Z 443 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:19.5502510Z     |
2019-08-26T13:42:19.5502909Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5502909Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5503008Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5503248Z 
2019-08-26T13:42:19.5503315Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5504121Z     |
2019-08-26T13:42:19.5504121Z     |
2019-08-26T13:42:19.5504372Z 447 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:19.5504499Z     |
2019-08-26T13:42:19.5504735Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5504735Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5504824Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5504868Z 
2019-08-26T13:42:19.5544197Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5553795Z    |
2019-08-26T13:42:19.5553795Z    |
2019-08-26T13:42:19.5557976Z 65 |     #[cfg_attr(feature = "capture", serde(skip))] //TODO
2019-08-26T13:42:19.5567921Z    |
2019-08-26T13:42:19.5573134Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5573134Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5576886Z    = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5580261Z 
2019-08-26T13:42:19.5613452Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5614007Z      |
2019-08-26T13:42:19.5614007Z      |
2019-08-26T13:42:19.5614067Z 2157 |     #[cfg_attr(feature = "capture", serde(skip))]
2019-08-26T13:42:19.5614227Z      |
2019-08-26T13:42:19.5614523Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5614523Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5614609Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5614676Z 
2019-08-26T13:42:19.5614749Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5615084Z      |
2019-08-26T13:42:19.5615084Z      |
2019-08-26T13:42:19.5615161Z 2219 |     #[cfg_attr(feature = "capture", serde(skip))] //TODO
2019-08-26T13:42:19.5615394Z      |
2019-08-26T13:42:19.5615657Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5615657Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5615765Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5615813Z 
2019-08-26T13:42:19.5770351Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5781473Z    |
2019-08-26T13:42:19.5781473Z    |
2019-08-26T13:42:19.5786338Z 60 |     #[cfg_attr(feature = "replay", serde(default = "FrameId::first"))]
2019-08-26T13:42:19.5795941Z    |
2019-08-26T13:42:19.5800873Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5800873Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.5805759Z    = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.5808404Z 
2019-08-26T13:42:19.5976158Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.5986186Z     |
2019-08-26T13:42:19.5986186Z     |
2019-08-26T13:42:19.5990648Z 496 |     #[cfg_attr(all(feature = "serde", any(feature = "capture", feature = "replay")), serde(skip))]
2019-08-26T13:42:19.5999240Z     |
2019-08-26T13:42:19.6004994Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.6004994Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.6009575Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.6012625Z 
2019-08-26T13:42:19.6043205Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.6048077Z     --> webrender/src/tiling.rs:1202:62
2019-08-26T13:42:19.6053529Z      |
2019-08-26T13:42:19.6057661Z 1202 |     #[cfg_attr(any(feature = "capture", feature = "replay"), serde(default = "FrameProfileCounters::new", skip))]
2019-08-26T13:42:19.6066923Z      |
2019-08-26T13:42:19.6072289Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.6072289Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.6076838Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:19.6079254Z 
2019-08-26T13:42:19.6138622Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:19.6139092Z     --> webrender/src/tiling.rs:1228:35
2019-08-26T13:42:19.6139783Z      |
2019-08-26T13:42:19.6139868Z 1228 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:19.6140035Z      |
2019-08-26T13:42:19.6140425Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.6140425Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:19.6140542Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:20.2244432Z warning: use of deprecated item 'try': use the `?` operator instead
2019-08-26T13:42:20.2249976Z    --> webrender/src/clip.rs:176:1
2019-08-26T13:42:20.2256365Z     |
2019-08-26T13:42:20.2261876Z 176 | / bitflags! {
---
2019-08-26T13:42:21.9654528Z     |
2019-08-26T13:42:21.9654672Z 713 |             let mut cbox: FT_BBox = mem::uninitialized();
2019-08-26T13:42:21.9654829Z     |                                     ^^^^^^^^^^^^^^^^^^
2019-08-26T13:42:21.9654963Z 
2019-08-26T13:42:35.8837508Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.8839300Z    --> webrender/src/frame_builder.rs:111:37
2019-08-26T13:42:35.8839688Z     |
2019-08-26T13:42:35.8839868Z 111 |     #[cfg_attr(feature = "capture", serde(skip))] //TODO
2019-08-26T13:42:35.8840244Z     |
2019-08-26T13:42:35.8840775Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8840775Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8840996Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.8841156Z 
2019-08-26T13:42:35.8877149Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.8878551Z     |
2019-08-26T13:42:35.8878551Z     |
2019-08-26T13:42:35.8878795Z 216 |     #[cfg_attr(any(feature = "capture", feature = "replay"), serde(skip))]
2019-08-26T13:42:35.8879175Z     |
2019-08-26T13:42:35.8879677Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8879677Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8879925Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.8880105Z 
2019-08-26T13:42:35.8923698Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.8924539Z     |
2019-08-26T13:42:35.8924539Z     |
2019-08-26T13:42:35.8924679Z 336 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:35.8924963Z     |
2019-08-26T13:42:35.8925532Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8925532Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8925717Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.8925830Z 
2019-08-26T13:42:35.8940428Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.8941703Z     |
2019-08-26T13:42:35.8941703Z     |
2019-08-26T13:42:35.8941841Z 443 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:35.8942117Z     |
2019-08-26T13:42:35.8942495Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8942495Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8942698Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.8942837Z 
2019-08-26T13:42:35.8947592Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.8948896Z     |
2019-08-26T13:42:35.8948896Z     |
2019-08-26T13:42:35.8949095Z 447 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:35.8949432Z     |
2019-08-26T13:42:35.8949859Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8949859Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.8950273Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.8950417Z 
2019-08-26T13:42:35.9063528Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9064480Z    |
2019-08-26T13:42:35.9064480Z    |
2019-08-26T13:42:35.9064620Z 65 |     #[cfg_attr(feature = "capture", serde(skip))] //TODO
2019-08-26T13:42:35.9064871Z    |
2019-08-26T13:42:35.9065266Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9065266Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9065438Z    = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.9065543Z 
2019-08-26T13:42:35.9091460Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9092935Z      |
2019-08-26T13:42:35.9092935Z      |
2019-08-26T13:42:35.9093091Z 2157 |     #[cfg_attr(feature = "capture", serde(skip))]
2019-08-26T13:42:35.9093656Z      |
2019-08-26T13:42:35.9094236Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9094236Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9094438Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.9094586Z 
2019-08-26T13:42:35.9101502Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9102588Z      |
2019-08-26T13:42:35.9102588Z      |
2019-08-26T13:42:35.9102891Z 2219 |     #[cfg_attr(feature = "capture", serde(skip))] //TODO
2019-08-26T13:42:35.9103420Z      |
2019-08-26T13:42:35.9104076Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9104076Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9104642Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.9104780Z 
2019-08-26T13:42:35.9316447Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9318271Z    |
2019-08-26T13:42:35.9318271Z    |
2019-08-26T13:42:35.9318506Z 60 |     #[cfg_attr(feature = "replay", serde(default = "FrameId::first"))]
2019-08-26T13:42:35.9318882Z    |
2019-08-26T13:42:35.9319430Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9319430Z    = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9319697Z    = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.9319858Z 
2019-08-26T13:42:35.9530161Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9531273Z     |
2019-08-26T13:42:35.9531273Z     |
2019-08-26T13:42:35.9531694Z 496 |     #[cfg_attr(all(feature = "serde", any(feature = "capture", feature = "replay")), serde(skip))]
2019-08-26T13:42:35.9532448Z     |
2019-08-26T13:42:35.9533316Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9533316Z     = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9533813Z     = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.9534111Z 
2019-08-26T13:42:35.9601915Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9602448Z     --> webrender/src/tiling.rs:1202:62
2019-08-26T13:42:35.9602816Z      |
2019-08-26T13:42:35.9602983Z 1202 |     #[cfg_attr(any(feature = "capture", feature = "replay"), serde(default = "FrameProfileCounters::new", skip))]
2019-08-26T13:42:35.9603579Z      |
2019-08-26T13:42:35.9603993Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9603993Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9604243Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:35.9604683Z 
2019-08-26T13:42:35.9611061Z error[E0658]: the attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future
2019-08-26T13:42:35.9611684Z     --> webrender/src/tiling.rs:1228:35
2019-08-26T13:42:35.9611747Z      |
2019-08-26T13:42:35.9611826Z 1228 |     #[cfg_attr(feature = "serde", serde(skip))]
2019-08-26T13:42:35.9611993Z      |
2019-08-26T13:42:35.9612271Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9612271Z      = note: for more information, see https://github.com/rust-lang/rust/issues/29642
2019-08-26T13:42:35.9612372Z      = help: add `#![feature(custom_attribute)]` to the crate attributes to enable
2019-08-26T13:42:36.5937293Z warning: use of deprecated item 'try': use the `?` operator instead
2019-08-26T13:42:36.5937670Z    --> webrender/src/clip.rs:176:1
2019-08-26T13:42:36.5938791Z     |
2019-08-26T13:42:36.5938880Z 176 | / bitflags! {
---
2019-08-26T13:42:38.4490856Z 
2019-08-26T13:42:44.9755895Z error: aborting due to 12 previous errors
2019-08-26T13:42:44.9756058Z 
2019-08-26T13:42:44.9757311Z For more information about this error, try `rustc --explain E0658`.
2019-08-26T13:42:45.1965990Z error: Could not compile `webrender`.
2019-08-26T13:43:02.2096952Z error: aborting due to 12 previous errors
2019-08-26T13:43:02.2097902Z 
2019-08-26T13:43:02.2099024Z For more information about this error, try `rustc --explain E0658`.
2019-08-26T13:43:02.2099024Z For more information about this error, try `rustc --explain E0658`.
2019-08-26T13:43:02.4469629Z error: Could not compile `webrender`.
2019-08-26T13:43:02.4470638Z To learn more, run the command again with --verbose.
2019-08-26T13:43:02.4496799Z thread 'main' panicked at 'tests failed for https://github.com/servo/webrender', src/tools/cargotest/main.rs:86:9
2019-08-26T13:43:02.4497392Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-08-26T13:43:02.4497518Z 
2019-08-26T13:43:02.4497518Z 
2019-08-26T13:43:02.4497557Z 
2019-08-26T13:43:02.4498118Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/cargotest" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build/ct"
2019-08-26T13:43:02.4498275Z expected success, got: exit code: 101
2019-08-26T13:43:02.4498323Z 
2019-08-26T13:43:02.4498374Z 
2019-08-26T13:43:02.4507400Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/test/run-fail/pretty src/test/run-pass-valgrind/pretty src/tools/cargo src/tools/cargotest
2019-08-26T13:43:02.4507568Z Build completed unsuccessfully in 2:06:56
2019-08-26T13:43:02.4556894Z Makefile:50: recipe for target 'check-aux' failed
2019-08-26T13:43:02.4557669Z make: *** [check-aux] Error 1
2019-08-26T13:43:02.4567592Z   local time: Mon Aug 26 13:43:02 UTC 2019
2019-08-26T13:43:02.5023206Z   network time: Mon, 26 Aug 2019 13:43:02 GMT
2019-08-26T13:43:02.5023410Z == end clock drift check ==
2019-08-26T13:43:02.5023410Z == end clock drift check ==
2019-08-26T13:43:04.0877747Z ##[error]Bash exited with code '2'.
2019-08-26T13:43:04.0923843Z ##[section]Starting: Upload CPU usage statistics
2019-08-26T13:43:04.0937893Z ==============================================================================
2019-08-26T13:43:04.0938012Z Task         : Bash
2019-08-26T13:43:04.0938100Z Description  : Run a Bash script on macOS, Linux, or Windows
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

--- a/src/librustc_metadata/decoder.rs
+++ b/src/librustc_metadata/decoder.rs
@@ -538,9 +538,7 @@ impl<'a, 'tcx> CrateMetadata {
                     attributes.iter().cloned().map(Symbol::intern).collect::<Vec<_>>();
                 (
                     trait_name,
-                    SyntaxExtensionKind::Derive(Box::new(ProcMacroDerive {
-                        client, attrs: helper_attrs.clone()
-                    })),
+                    SyntaxExtensionKind::Derive(Box::new(ProcMacroDerive { client })),
                     helper_attrs,
                 )
             }

--- a/src/librustc_resolve/macros.rs
+++ b/src/librustc_resolve/macros.rs
@@ -13,7 +13,7 @@ use rustc::{ty, lint, span_bug};
 use syntax::ast::{self, NodeId, Ident};
 use syntax::attr::StabilityLevel;
 use syntax::edition::Edition;
-use syntax::ext::base::{self, Indeterminate, SpecialDerives};
+use syntax::ext::base::{self, InvocationRes, Indeterminate, SpecialDerives};
 use syntax::ext::base::{MacroKind, SyntaxExtension};
 use syntax::ext::expand::{AstFragment, Invocation, InvocationKind};
 use syntax::ext::hygiene::{self, ExpnId, ExpnData, ExpnKind};
@@ -142,7 +142,7 @@ impl<'a> base::Resolver for Resolver<'a> {
 
     fn resolve_macro_invocation(
         &mut self, invoc: &Invocation, eager_expansion_root: ExpnId, force: bool
-    ) -> Result<Option<Lrc<SyntaxExtension>>, Indeterminate> {
+    ) -> Result<InvocationRes, Indeterminate> {
         let invoc_id = invoc.expansion_data.id;
         let parent_scope = match self.invocation_parent_scopes.get(&invoc_id) {
             Some(parent_scope) => *parent_scope,
@@ -165,25 +165,24 @@ impl<'a> base::Resolver for Resolver<'a> {
             InvocationKind::Derive { ref path, .. } =>
                 (path, MacroKind::Derive, &[][..], false),
             InvocationKind::DeriveContainer { ref derives, .. } => {
-                // Block expansion of derives in the container until we know whether one of them
-                // is a built-in `Copy`. Skip the resolution if there's only one derive - either
-                // it's not a `Copy` and we don't need to do anything, or it's a `Copy` and it
-                // will automatically knows about itself.
-                let mut result = Ok(None);
-                if derives.len() > 1 {
-                    for path in derives {
-                        match self.resolve_macro_path(path, Some(MacroKind::Derive),
-                                                      &parent_scope, true, force) {
-                            Ok((Some(ref ext), _)) if ext.is_derive_copy => {
-                                self.add_derives(invoc_id, SpecialDerives::COPY);
-                                return Ok(None);
-                            }
-                            Err(Determinacy::Undetermined) => result = Err(Indeterminate),
-                            _ => {}
-                        }
-                    }
+                // Block expansion of the container until we resolve all derives in it.
+                // This is required for two reasons:
+                // - Derive helper attributes are in scope for the item to which the `#[derive]`
+                //   is applied, so they have to be produced by the container's expansion rather
+                //   than by individual derives.
+                // - Derives in the container need to know whether one of them is a built-in `Copy`.
+                // FIXME: Try to avoid repeated resolutions for derives here and in expansion.
+                let mut exts = Vec::new();
+                for path in derives {
+                    exts.push(match self.resolve_macro_path(
+                        path, Some(MacroKind::Derive), &parent_scope, true, force
+                    ) {
+                        Ok((Some(ext), _)) => ext,
+                        Ok(_) | Err(Determinacy::Determined) => self.dummy_ext(MacroKind::Derive),
+                        Err(Determinacy::Undetermined) => return Err(Indeterminate),
+                    })
                 }
-                return result;
+                return Ok(InvocationRes::DeriveContainer(exts));
             }
         };
 
@@ -203,7 +202,7 @@ impl<'a> base::Resolver for Resolver<'a> {
             self.definitions.add_parent_module_of_macro_def(invoc_id, normal_module_def_id);
         }
 
-        Ok(Some(ext))
+        Ok(InvocationRes::Single(ext))
     }
 
     fn check_unused_macros(&self) {

--- a/src/libsyntax/ext/base.rs
+++ b/src/libsyntax/ext/base.rs
@@ -11,6 +11,7 @@ use crate::ptr::P;
 use crate::symbol::{kw, sym, Ident, Symbol};
 use crate::{ThinVec, MACRO_ARGUMENTS};
 use crate::tokenstream::{self, TokenStream, TokenTree};
+use crate::visit::Visitor;
 
 use errors::{DiagnosticBuilder, DiagnosticId};
 use smallvec::{smallvec, SmallVec};
@@ -69,6 +70,17 @@ impl Annotatable {
             Annotatable::ForeignItem(ref foreign_item) => foreign_item.span,
             Annotatable::Stmt(ref stmt) => stmt.span,
             Annotatable::Expr(ref expr) => expr.span,
+        }
+    }
+
+    pub fn visit_with<'a, V: Visitor<'a>>(&'a self, visitor: &mut V) {
+        match self {
+            Annotatable::Item(item) => visitor.visit_item(item),
+            Annotatable::TraitItem(trait_item) => visitor.visit_trait_item(trait_item),
+            Annotatable::ImplItem(impl_item) => visitor.visit_impl_item(impl_item),
+            Annotatable::ForeignItem(foreign_item) => visitor.visit_foreign_item(foreign_item),
+            Annotatable::Stmt(stmt) => visitor.visit_stmt(stmt),
+            Annotatable::Expr(expr) => visitor.visit_expr(expr),
         }
     }
 
@@ -637,6 +649,12 @@ impl SyntaxExtension {
 
 pub type NamedSyntaxExtension = (Name, SyntaxExtension);
 
+/// Result of resolving a macro invocation.
+pub enum InvocationRes {
+    Single(Lrc<SyntaxExtension>),
+    DeriveContainer(Vec<Lrc<SyntaxExtension>>),
+}
+
 /// Error type that denotes indeterminacy.
 pub struct Indeterminate;
 
@@ -664,7 +682,7 @@ pub trait Resolver {
 
     fn resolve_macro_invocation(
         &mut self, invoc: &Invocation, eager_expansion_root: ExpnId, force: bool
-    ) -> Result<Option<Lrc<SyntaxExtension>>, Indeterminate>;
+    ) -> Result<InvocationRes, Indeterminate>;
 
     fn check_unused_macros(&self);
 

--- a/src/libsyntax/ext/expand.rs
+++ b/src/libsyntax/ext/expand.rs
@@ -4,7 +4,7 @@ use crate::attr::{self, HasAttrs};
 use crate::source_map::respan;
 use crate::config::StripUnconfigured;
 use crate::ext::base::*;
-use crate::ext::proc_macro::collect_derives;
+use crate::ext::proc_macro::{collect_derives, MarkAttrs};
 use crate::ext::hygiene::{ExpnId, SyntaxContext, ExpnData, ExpnKind};
 use crate::ext::tt::macro_rules::annotate_err_with_kind;
 use crate::ext::placeholders::{placeholder, PlaceholderExpander};
@@ -307,10 +307,10 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
 
             let eager_expansion_root =
                 if self.monotonic { invoc.expansion_data.id } else { orig_expansion_data.id };
-            let ext = match self.cx.resolver.resolve_macro_invocation(
+            let res = match self.cx.resolver.resolve_macro_invocation(
                 &invoc, eager_expansion_root, force
             ) {
-                Ok(ext) => ext,
+                Ok(res) => res,
                 Err(Indeterminate) => {
                     undetermined_invocations.push(invoc);
                     continue
@@ -322,54 +322,72 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
             self.cx.current_expansion = invoc.expansion_data.clone();
 
             // FIXME(jseyfried): Refactor out the following logic
-            let (expanded_fragment, new_invocations) = if let Some(ext) = ext {
-                let fragment = self.expand_invoc(invoc, &ext.kind);
-                self.collect_invocations(fragment, &[])
-            } else if let InvocationKind::DeriveContainer { derives: traits, item } = invoc.kind {
-                if !item.derive_allowed() {
-                    let attr = attr::find_by_name(item.attrs(), sym::derive)
-                        .expect("`derive` attribute should exist");
-                    let span = attr.span;
-                    let mut err = self.cx.mut_span_err(span,
-                                                        "`derive` may only be applied to \
-                                                        structs, enums and unions");
-                    if let ast::AttrStyle::Inner = attr.style {
-                        let trait_list = traits.iter()
-                            .map(|t| t.to_string()).collect::<Vec<_>>();
-                        let suggestion = format!("#[derive({})]", trait_list.join(", "));
-                        err.span_suggestion(
-                            span, "try an outer attribute", suggestion,
-                            // We don't 𝑘𝑛𝑜𝑤 that the following item is an ADT
-                            Applicability::MaybeIncorrect
-                        );
+            let (expanded_fragment, new_invocations) = match res {
+                InvocationRes::Single(ext) => {
+                    let fragment = self.expand_invoc(invoc, &ext.kind);
+                    self.collect_invocations(fragment, &[])
+                }
+                InvocationRes::DeriveContainer(exts) => {
+                    let (derives, item) = match invoc.kind {
+                        InvocationKind::DeriveContainer { derives, item } => (derives, item),
+                        _ => unreachable!(),
+                    };
+                    if !item.derive_allowed() {
+                        let attr = attr::find_by_name(item.attrs(), sym::derive)
+                            .expect("`derive` attribute should exist");
+                        let span = attr.span;
+                        let mut err = self.cx.mut_span_err(span,
+                            "`derive` may only be applied to structs, enums and unions");
+                        if let ast::AttrStyle::Inner = attr.style {
+                            let trait_list = derives.iter()
+                                .map(|t| t.to_string()).collect::<Vec<_>>();
+                            let suggestion = format!("#[derive({})]", trait_list.join(", "));
+                            err.span_suggestion(
+                                span, "try an outer attribute", suggestion,
+                                // We don't 𝑘𝑛𝑜𝑤 that the following item is an ADT
+                                Applicability::MaybeIncorrect
+                            );
+                        }
+                        err.emit();
                     }
-                    err.emit();
-                }
 
-                let mut item = self.fully_configure(item);
-                item.visit_attrs(|attrs| attrs.retain(|a| a.path != sym::derive));
-                let derive_placeholders =
-                    all_derive_placeholders.entry(invoc.expansion_data.id).or_default();
+                    let mut item = self.fully_configure(item);
+                    item.visit_attrs(|attrs| attrs.retain(|a| a.path != sym::derive));
+                    let mut helper_attrs = Vec::new();
+                    let mut has_copy = false;
+                    for ext in exts {
+                        helper_attrs.extend(&ext.helper_attrs);
+                        has_copy |= ext.is_derive_copy;
+                    }
+                    // Mark derive helpers inside this item as known and used.
+                    // FIXME: This is a hack, derive helpers should be integrated with regular name
+                    // resolution instead. For example, helpers introduced by a derive container
+                    // can be in scope for all code produced by that container's expansion.
+                    item.visit_with(&mut MarkAttrs(&helper_attrs));
+                    if has_copy {
+                        self.cx.resolver.add_derives(invoc.expansion_data.id, SpecialDerives::COPY);
+                    }
 
-                derive_placeholders.reserve(traits.len());
-                invocations.reserve(traits.len());
-                for path in traits {
-                    let expn_id = ExpnId::fresh(None);
-                    derive_placeholders.push(NodeId::placeholder_from_expn_id(expn_id));
-                    invocations.push(Invocation {
-                        kind: InvocationKind::Derive { path, item: item.clone() },
-                        fragment_kind: invoc.fragment_kind,
-                        expansion_data: ExpansionData {
-                            id: expn_id,
-                            ..invoc.expansion_data.clone()
-                        },
-                    });
+                    let derive_placeholders =
+                        all_derive_placeholders.entry(invoc.expansion_data.id).or_default();
+                    derive_placeholders.reserve(derives.len());
+                    invocations.reserve(derives.len());
+                    for path in derives {
+                        let expn_id = ExpnId::fresh(None);
+                        derive_placeholders.push(NodeId::placeholder_from_expn_id(expn_id));
+                        invocations.push(Invocation {
+                            kind: InvocationKind::Derive { path, item: item.clone() },
+                            fragment_kind: invoc.fragment_kind,
+                            expansion_data: ExpansionData {
+                                id: expn_id,
+                                ..invoc.expansion_data.clone()
+                            },
+                        });
+                    }
+                    let fragment = invoc.fragment_kind
+                        .expect_from_annotatables(::std::iter::once(item));
+                    self.collect_invocations(fragment, derive_placeholders)
                 }
-                let fragment = invoc.fragment_kind
-                    .expect_from_annotatables(::std::iter::once(item));
-                self.collect_invocations(fragment, derive_placeholders)
-            } else {
-                unreachable!()
             };
 
             if expanded_fragments.len() < depth {

--- a/src/libsyntax/ext/proc_macro.rs
+++ b/src/libsyntax/ext/proc_macro.rs
@@ -78,7 +78,6 @@ pub struct ProcMacroDerive {
     pub client: proc_macro::bridge::client::Client<
         fn(proc_macro::TokenStream) -> proc_macro::TokenStream,
     >,
-    pub attrs: Vec<ast::Name>,
 }
 
 impl MultiItemModifier for ProcMacroDerive {
@@ -110,9 +109,6 @@ impl MultiItemModifier for ProcMacroDerive {
                 return Vec::new()
             }
         }
-
-        // Mark attributes as known, and used.
-        MarkAttrs(&self.attrs).visit_item(&item);
 
         let token = token::Interpolated(Lrc::new(token::NtItem(item)));
         let input = tokenstream::TokenTree::token(token, DUMMY_SP).into();
@@ -164,7 +160,7 @@ impl MultiItemModifier for ProcMacroDerive {
     }
 }
 
-struct MarkAttrs<'a>(&'a [ast::Name]);
+crate struct MarkAttrs<'a>(crate &'a [ast::Name]);
 
 impl<'a> Visitor<'a> for MarkAttrs<'a> {
     fn visit_attribute(&mut self, attr: &Attribute) {

--- a/src/test/ui/derives/deriving-bounds.stderr
+++ b/src/test/ui/derives/deriving-bounds.stderr
@@ -1,15 +1,3 @@
-error: cannot find derive macro `Send` in this scope
-  --> $DIR/deriving-bounds.rs:1:10
-   |
-LL | #[derive(Send)]
-   |          ^^^^
-   |
-note: unsafe traits like `Send` should be implemented explicitly
-  --> $DIR/deriving-bounds.rs:1:10
-   |
-LL | #[derive(Send)]
-   |          ^^^^
-
 error: cannot find derive macro `Sync` in this scope
   --> $DIR/deriving-bounds.rs:5:10
    |
@@ -20,6 +8,18 @@ note: unsafe traits like `Sync` should be implemented explicitly
   --> $DIR/deriving-bounds.rs:5:10
    |
 LL | #[derive(Sync)]
+   |          ^^^^
+
+error: cannot find derive macro `Send` in this scope
+  --> $DIR/deriving-bounds.rs:1:10
+   |
+LL | #[derive(Send)]
+   |          ^^^^
+   |
+note: unsafe traits like `Send` should be implemented explicitly
+  --> $DIR/deriving-bounds.rs:1:10
+   |
+LL | #[derive(Send)]
    |          ^^^^
 
 error: aborting due to 2 previous errors

--- a/src/test/ui/feature-gate/issue-43106-gating-of-derive-2.stderr
+++ b/src/test/ui/feature-gate/issue-43106-gating-of-derive-2.stderr
@@ -1,5 +1,5 @@
 error: cannot find derive macro `x3300` in this scope
-  --> $DIR/issue-43106-gating-of-derive-2.rs:4:14
+  --> $DIR/issue-43106-gating-of-derive-2.rs:12:14
    |
 LL |     #[derive(x3300)]
    |              ^^^^^
@@ -11,7 +11,7 @@ LL |     #[derive(x3300)]
    |              ^^^^^
 
 error: cannot find derive macro `x3300` in this scope
-  --> $DIR/issue-43106-gating-of-derive-2.rs:12:14
+  --> $DIR/issue-43106-gating-of-derive-2.rs:4:14
    |
 LL |     #[derive(x3300)]
    |              ^^^^^

--- a/src/test/ui/feature-gate/issue-43106-gating-of-derive.rs
+++ b/src/test/ui/feature-gate/issue-43106-gating-of-derive.rs
@@ -1,9 +1,6 @@
 // `#![derive]` raises errors when it occurs at contexts other than ADT
 // definitions.
 
-#![derive(Debug)]
-//~^ ERROR `derive` may only be applied to structs, enums and unions
-
 #[derive(Debug)]
 //~^ ERROR `derive` may only be applied to structs, enums and unions
 mod derive {

--- a/src/test/ui/feature-gate/issue-43106-gating-of-derive.stderr
+++ b/src/test/ui/feature-gate/issue-43106-gating-of-derive.stderr
@@ -1,38 +1,32 @@
 error: `derive` may only be applied to structs, enums and unions
   --> $DIR/issue-43106-gating-of-derive.rs:4:1
    |
-LL | #![derive(Debug)]
-   | ^^^^^^^^^^^^^^^^^ help: try an outer attribute: `#[derive(Debug)]`
-
-error: `derive` may only be applied to structs, enums and unions
-  --> $DIR/issue-43106-gating-of-derive.rs:7:1
-   |
 LL | #[derive(Debug)]
    | ^^^^^^^^^^^^^^^^
 
 error: `derive` may only be applied to structs, enums and unions
-  --> $DIR/issue-43106-gating-of-derive.rs:10:17
+  --> $DIR/issue-43106-gating-of-derive.rs:7:17
    |
 LL |     mod inner { #![derive(Debug)] }
    |                 ^^^^^^^^^^^^^^^^^ help: try an outer attribute: `#[derive(Debug)]`
 
 error: `derive` may only be applied to structs, enums and unions
-  --> $DIR/issue-43106-gating-of-derive.rs:13:5
+  --> $DIR/issue-43106-gating-of-derive.rs:10:5
    |
 LL |     #[derive(Debug)]
    |     ^^^^^^^^^^^^^^^^
 
 error: `derive` may only be applied to structs, enums and unions
-  --> $DIR/issue-43106-gating-of-derive.rs:26:5
+  --> $DIR/issue-43106-gating-of-derive.rs:23:5
    |
 LL |     #[derive(Debug)]
    |     ^^^^^^^^^^^^^^^^
 
 error: `derive` may only be applied to structs, enums and unions
-  --> $DIR/issue-43106-gating-of-derive.rs:30:5
+  --> $DIR/issue-43106-gating-of-derive.rs:27:5
    |
 LL |     #[derive(Debug)]
    |     ^^^^^^^^^^^^^^^^
 
-error: aborting due to 6 previous errors
+error: aborting due to 5 previous errors
 

--- a/src/test/ui/issues/issue-36617.rs
+++ b/src/test/ui/issues/issue-36617.rs
@@ -1,3 +1,4 @@
 #![derive(Copy)] //~ ERROR `derive` may only be applied to structs, enums and unions
+                 //~| ERROR cannot determine resolution for the derive macro `Copy`
 
 fn main() {}

--- a/src/test/ui/issues/issue-36617.stderr
+++ b/src/test/ui/issues/issue-36617.stderr
@@ -4,5 +4,13 @@ error: `derive` may only be applied to structs, enums and unions
 LL | #![derive(Copy)]
    | ^^^^^^^^^^^^^^^^ help: try an outer attribute: `#[derive(Copy)]`
 
-error: aborting due to previous error
+error: cannot determine resolution for the derive macro `Copy`
+  --> $DIR/issue-36617.rs:1:11
+   |
+LL | #![derive(Copy)]
+   |           ^^^^
+   |
+   = note: import resolution is stuck, try simplifying macro imports
+
+error: aborting due to 2 previous errors
 

--- a/src/test/ui/proc-macro/derive-helper-configured.rs
+++ b/src/test/ui/proc-macro/derive-helper-configured.rs
@@ -1,0 +1,18 @@
+// Derive helpers are resolved successfully inside `cfg_attr`.
+
+// check-pass
+// compile-flats:--cfg TRUE
+// aux-build:test-macros.rs
+
+#[macro_use]
+extern crate test_macros;
+
+#[cfg_attr(TRUE, empty_helper)]
+#[derive(Empty)]
+#[cfg_attr(TRUE, empty_helper)]
+struct S {
+    #[cfg_attr(TRUE, empty_helper)]
+    field: u8,
+}
+
+fn main() {}

--- a/src/test/ui/proc-macro/derive-helper-shadowing.rs
+++ b/src/test/ui/proc-macro/derive-helper-shadowing.rs
@@ -19,7 +19,8 @@ struct S {
         struct U;
 
         mod inner {
-            #[empty_helper] //~ ERROR cannot find attribute macro `empty_helper` in this scope
+            // FIXME No ambiguity, attributes in non-macro positions are not resolved properly
+            #[empty_helper]
             struct V;
         }
 

--- a/src/test/ui/proc-macro/derive-helper-shadowing.stderr
+++ b/src/test/ui/proc-macro/derive-helper-shadowing.stderr
@@ -1,9 +1,3 @@
-error: cannot find attribute macro `empty_helper` in this scope
-  --> $DIR/derive-helper-shadowing.rs:22:15
-   |
-LL |             #[empty_helper]
-   |               ^^^^^^^^^^^^
-
 error[E0659]: `empty_helper` is ambiguous (derive helper attribute vs any other name)
   --> $DIR/derive-helper-shadowing.rs:8:3
    |
@@ -22,6 +16,6 @@ LL | use test_macros::empty_attr as empty_helper;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: use `crate::empty_helper` to refer to this attribute macro unambiguously
 
-error: aborting due to 2 previous errors
+error: aborting due to previous error
 
 For more information about this error, try `rustc --explain E0659`.

--- a/src/test/ui/proc-macro/resolve-error.stderr
+++ b/src/test/ui/proc-macro/resolve-error.stderr
@@ -1,50 +1,8 @@
-error: cannot find derive macro `FooWithLongNan` in this scope
-  --> $DIR/resolve-error.rs:22:10
+error: cannot find macro `bang_proc_macrp!` in this scope
+  --> $DIR/resolve-error.rs:56:5
    |
-LL | #[derive(FooWithLongNan)]
-   |          ^^^^^^^^^^^^^^ help: a derive macro with a similar name exists: `FooWithLongName`
-
-error: cannot find attribute macro `attr_proc_macra` in this scope
-  --> $DIR/resolve-error.rs:27:3
-   |
-LL | #[attr_proc_macra]
-   |   ^^^^^^^^^^^^^^^ help: an attribute macro with a similar name exists: `attr_proc_macro`
-
-error: cannot find attribute macro `FooWithLongNan` in this scope
-  --> $DIR/resolve-error.rs:31:3
-   |
-LL | #[FooWithLongNan]
-   |   ^^^^^^^^^^^^^^
-
-error: cannot find derive macro `Dlone` in this scope
-  --> $DIR/resolve-error.rs:34:10
-   |
-LL | #[derive(Dlone)]
-   |          ^^^^^ help: a derive macro with a similar name exists: `Clone`
-
-error: cannot find derive macro `Dlona` in this scope
-  --> $DIR/resolve-error.rs:38:10
-   |
-LL | #[derive(Dlona)]
-   |          ^^^^^ help: a derive macro with a similar name exists: `Clona`
-
-error: cannot find derive macro `attr_proc_macra` in this scope
-  --> $DIR/resolve-error.rs:42:10
-   |
-LL | #[derive(attr_proc_macra)]
-   |          ^^^^^^^^^^^^^^^
-
-error: cannot find macro `FooWithLongNama!` in this scope
-  --> $DIR/resolve-error.rs:47:5
-   |
-LL |     FooWithLongNama!();
-   |     ^^^^^^^^^^^^^^^ help: a macro with a similar name exists: `FooWithLongNam`
-
-error: cannot find macro `attr_proc_macra!` in this scope
-  --> $DIR/resolve-error.rs:50:5
-   |
-LL |     attr_proc_macra!();
-   |     ^^^^^^^^^^^^^^^ help: a macro with a similar name exists: `attr_proc_mac`
+LL |     bang_proc_macrp!();
+   |     ^^^^^^^^^^^^^^^ help: a macro with a similar name exists: `bang_proc_macro`
 
 error: cannot find macro `Dlona!` in this scope
   --> $DIR/resolve-error.rs:53:5
@@ -52,11 +10,53 @@ error: cannot find macro `Dlona!` in this scope
 LL |     Dlona!();
    |     ^^^^^
 
-error: cannot find macro `bang_proc_macrp!` in this scope
-  --> $DIR/resolve-error.rs:56:5
+error: cannot find macro `attr_proc_macra!` in this scope
+  --> $DIR/resolve-error.rs:50:5
    |
-LL |     bang_proc_macrp!();
-   |     ^^^^^^^^^^^^^^^ help: a macro with a similar name exists: `bang_proc_macro`
+LL |     attr_proc_macra!();
+   |     ^^^^^^^^^^^^^^^ help: a macro with a similar name exists: `attr_proc_mac`
+
+error: cannot find macro `FooWithLongNama!` in this scope
+  --> $DIR/resolve-error.rs:47:5
+   |
+LL |     FooWithLongNama!();
+   |     ^^^^^^^^^^^^^^^ help: a macro with a similar name exists: `FooWithLongNam`
+
+error: cannot find derive macro `attr_proc_macra` in this scope
+  --> $DIR/resolve-error.rs:42:10
+   |
+LL | #[derive(attr_proc_macra)]
+   |          ^^^^^^^^^^^^^^^
+
+error: cannot find derive macro `Dlona` in this scope
+  --> $DIR/resolve-error.rs:38:10
+   |
+LL | #[derive(Dlona)]
+   |          ^^^^^ help: a derive macro with a similar name exists: `Clona`
+
+error: cannot find derive macro `Dlone` in this scope
+  --> $DIR/resolve-error.rs:34:10
+   |
+LL | #[derive(Dlone)]
+   |          ^^^^^ help: a derive macro with a similar name exists: `Clone`
+
+error: cannot find attribute macro `FooWithLongNan` in this scope
+  --> $DIR/resolve-error.rs:31:3
+   |
+LL | #[FooWithLongNan]
+   |   ^^^^^^^^^^^^^^
+
+error: cannot find attribute macro `attr_proc_macra` in this scope
+  --> $DIR/resolve-error.rs:27:3
+   |
+LL | #[attr_proc_macra]
+   |   ^^^^^^^^^^^^^^^ help: an attribute macro with a similar name exists: `attr_proc_macro`
+
+error: cannot find derive macro `FooWithLongNan` in this scope
+  --> $DIR/resolve-error.rs:22:10
+   |
+LL | #[derive(FooWithLongNan)]
+   |          ^^^^^^^^^^^^^^ help: a derive macro with a similar name exists: `FooWithLongName`
 
 error: aborting due to 10 previous errors
 

